File: client-apps/cli/internal/cli/apply/runtime.go
// Package apply provides SDK synthesis execution for the stigmer apply command.
package apply

import (
	"fmt"
	"path/filepath"
	"strings"
)

// Runtime identifies the language runtime used to execute an SDK entry point.
//
// This is a CLI-local type that replaces the removed ProjectRuntime proto enum.
// The runtime is inferred from the entry_point file extension rather than being
// declared explicitly by the user.
type Runtime string

const (
	RuntimeGo     Runtime = "go"
	RuntimePython Runtime = "python"
	RuntimeNode   Runtime = "node"
)

// extensionToRuntime maps file extensions to their SDK runtime.
// Keep in sync with supportedEntryPointExtensions in project/validator.go.
var extensionToRuntime = map[string]Runtime{
	".go":  RuntimeGo,
	".py":  RuntimePython,
	".ts":  RuntimeNode,
	".js":  RuntimeNode,
	".mts": RuntimeNode,
	".mjs": RuntimeNode,
}

// InferRuntime determines the SDK runtime from an entry point's file extension.
// Returns an error with supported extensions listed if the extension is unrecognized.
func InferRuntime(entryPoint string) (Runtime, error) {
	ext := strings.ToLower(filepath.Ext(entryPoint))
	if rt, ok := extensionToRuntime[ext]; ok {
		return rt, nil
	}

	supported := make([]string, 0, len(extensionToRuntime))
	for ext := range extensionToRuntime {
		supported = append(supported, ext)
	}

	return "", fmt.Errorf(
		"cannot infer runtime from entry point %q (extension %q)\n\n"+
			"Supported extensions: %s",
		entryPoint, ext, strings.Join(supported, ", "),
	)
}

File: client-apps/cli-ts/src/resources/apply/synth/runtime.ts
// Runtime inference for SDK project synthesis.
//
// A project's `entry_point` extension determines which language runtime executes
// it — there is no explicit `runtime` field (the proto enum was removed; the
// extension is the single source of truth). Port of the Go CLI's
// `internal/cli/apply/runtime.go`; the extension table is kept in lockstep with
// the backend's `supportedEntryPointExtensions` (project validator).

import { extname } from "node:path";
import { UsageError } from "../../../errors/index.js";

/** Language runtime used to execute an SDK entry point. */
export type Runtime = "go" | "python" | "node";

// Extension → runtime. Keep in sync with the backend project validator and Go's
// extensionToRuntime.
const EXTENSION_TO_RUNTIME: ReadonlyMap<string, Runtime> = new Map([
  [".go", "go"],
  [".py", "python"],
  [".ts", "node"],
  [".js", "node"],
  [".mts", "node"],
  [".mjs", "node"],
]);

/**
 * Infer the runtime from an entry point's file extension. Throws a UsageError
 * listing the supported extensions when the extension is unrecognized.
 */
export function inferRuntime(entryPoint: string): Runtime {
  const ext = extname(entryPoint).toLowerCase();
  const runtime = EXTENSION_TO_RUNTIME.get(ext);
  if (runtime !== undefined) return runtime;

  const supported = [...EXTENSION_TO_RUNTIME.keys()].join(", ");
  throw new UsageError(
    `cannot infer runtime from entry point '${entryPoint}' (extension '${ext}')\n\n` + `Supported extensions: ${supported}`,
  );
}
